providers/datastore/shedlock-provider-datastore/src/main/java/net/javacrumbs/shedlock/provider/datastore/DatastoreStorageAccessor.java

```diff
@@ -0,0 +1,168 @@
+import org.slf4j.LoggerFactory;
+
+import java.time.Instant;
+import java.util.Optional;
+import java.util.function.Function;
+
+import static java.util.Objects.requireNonNull;
+import static java.util.Optional.ofNullable;
+
+public class DatastoreStorageAccessor extends AbstractStorageAccessor {
+    private static final Logger log = LoggerFactory.getLogger(DatastoreStorageAccessor.class);
+
+    private final Datastore datastore;
+    private final String hostname;
+    private final String entityName;
+    private final DatastoreLockProvider.Fields fields;
+
+    public DatastoreStorageAccessor(DatastoreLockProvider.Configuration configuration) {
+        requireNonNull(configuration);
+        this.datastore = configuration.getDatastore();
+        this.hostname = Utils.getHostname();
+        this.entityName = configuration.getEntityName();
+        this.fields = configuration.getFields();
+    }
+
+    @Override
+    public boolean insertRecord(LockConfiguration config) {
+        return insert(config.getName(), config.getLockAtMostUntil());
+    }
+
+    @Override
+    public boolean updateRecord(LockConfiguration config) {
+        return updateExisting(config.getName(), config.getLockAtMostUntil());
+    }
+
+    @Override
+    public void unlock(LockConfiguration config) {
+        updateOwn(config.getName(), config.getUnlockTime());
+    }
+
+    @Override
+    public boolean extend(LockConfiguration config) {
+        return updateOwn(config.getName(), config.getLockAtMostUntil());
+    }
+
+    private boolean insert(String name, Instant until) {
+        return doInTxn(txn -> {
+            KeyFactory keyFactory = this.datastore.newKeyFactory().setKind(this.entityName);
+            Key key = keyFactory.newKey(name);
+            Entity entity = Entity.newBuilder(key)
+                .set(this.fields.lockUntil(), fromInstant(until))
+                .set(this.fields.lockedAt(), fromInstant(ClockProvider.now()))
+                .set(this.fields.lockedBy(), this.hostname)
+                .build();
+            txn.add(entity);
+            txn.commit();
+            return Optional.of(true);
+        }).orElse(false);
+    }
+
+    private boolean updateExisting(String name, Instant until) {
+        return doInTxn(txn ->
+            get(name, txn)
+                .filter(entity -> {
+                    var now = ClockProvider.now();
+                    var lockUntilTs = nullableTimestamp(entity, this.fields.lockUntil());
+                    return lockUntilTs != null && lockUntilTs.isBefore(now);
+                })
+                .map(entity -> {
+                    txn.put(Entity.newBuilder(entity)
+                        .set(this.fields.lockUntil(), fromInstant(until))
+                        .set(this.fields.lockedAt(), fromInstant(ClockProvider.now()))
+                        .set(this.fields.lockedBy(), this.hostname)
+                        .build());
+                    txn.commit();
```

Comment:
Wouldn't it be better to commit it in `doInTxn`?

Reply:
I had a different idea in the beginning; but given the final product it makes very little sense to have it there. Will have the commit in the txn to reflect the now observed behaviour.